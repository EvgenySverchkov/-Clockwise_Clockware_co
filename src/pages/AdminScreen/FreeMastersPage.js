import React from "react";
import { useDispatch, useSelector } from "react-redux";

import FreeMastersForm from "../../forms/FreeMastersForm";

import { addCurrentOrderToState } from "../../store/adminPanel/orders/actions";
import {
  changeMasterListIsLoad,
  changeSuccessModalDataAdmin,
  changeModalWarningDataAdmin,
} from "../../store/adminPanel/services/actions";

import { SERVERDOMAIN } from "../../services/serverUrls";
import sendConfirmEmail from "../../services/mailSendler";

function MastersList(props) {
  const state = useSelector((state) => {
    return {
      suitableMasters: state.master_reducer.suitableMasters,
      masterListIsLoad: state.main_adminPanel_reduser.masterListIsLoad,
    };
  });
  const dispatch = useDispatch();

  function submitHandler(e) {
    e.preventDefault();
    let masterId = e.target.chooseMaster.value;
    if (!e.target.chooseMaster.length) {
      if (!e.target.chooseMaster.checked) {
        masterId = null;
      }
    }
    if (!masterId) {
      dispatch(changeModalWarningDataAdmin({ msg: "Please, choose one!!!" }));
      document.getElementById("callWarningModalBtn").click();
      return false;
    }
    let endOrderTime;
    let clientTimeHour = +props.currentOrder.time.match(/[^:]+/);
    let clientTimeMin = props.currentOrder.time.match(/:\d\d$/);
    switch (props.currentOrder.size) {
      case "small":
        endOrderTime = clientTimeHour + 1 + clientTimeMin;
        break;
      case "middle":
        endOrderTime = clientTimeHour + 2 + clientTimeMin;
        break;
      case "large":
        endOrderTime = clientTimeHour + 3 + clientTimeMin;
        break;
      default:
        endOrderTime = 0;
    }
    let newObj = {
      ...props.currentOrder,
      masterId: masterId,
      endTime: endOrderTime,
    };

    dispatch(changeMasterListIsLoad(true));
    fetch(`${SERVERDOMAIN}/orders/post`, {
      method: "POST",
      headers: {
        Authorization: sessionStorage.getItem("token")
          ? "Bearer " + sessionStorage.getItem("token")
          : "",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(newObj),
    })
      .then((json) => json.json())
      .then((data) => {
        dispatch(changeMasterListIsLoad(false));
        if (data.success) {
          dispatch(
            changeSuccessModalDataAdmin({
              msg: data.msg,
              backBtnTxt: "Go to the list of orders",
              backTo: "/admin/ordersList",
            })
          );
          props.history.push("/admin");
          document.getElementById("callSuccessModalBtn").click();
          dispatch(addCurrentOrderToState({}));
          sendConfirmEmail(`${SERVERDOMAIN}/send_message`, data.payload);
        } else {
          dispatch(changeModalWarningDataAdmin({ msg: data.msg }));
          document.getElementById("callWarningModalBtn").click();
        }
      })
      .catch((err) => {
        dispatch(changeMasterListIsLoad(false));
        alert(err);
      });
  }
  return (
    <FreeMastersForm
      submitHandler={submitHandler}
      isLoad={state.masterListIsLoad}
      suitableMasters={state.suitableMasters}
      backTo={"/admin/addOrderForm"}
    />
  );
}

export default MastersList;
