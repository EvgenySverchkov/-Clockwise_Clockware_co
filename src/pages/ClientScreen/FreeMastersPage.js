import React from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";
import { addCurrentOrderToState } from "../../store/clientSide/data/actions";
import {
  changeMasterListIsLoad,
  changeSuccessModalData,
  changeWarningModalData,
} from "../../store/clientSide/services/actions";
import FreeMastersForm from "../../forms/FreeMastersForm";
import sendMail from "../../services/mailSendler";

import { SERVERDOMAIN } from "../../services/serverUrls";
function MastersList() {
  const state = useSelector((state) => {
    return {
      suitableMasters: state.clientOrderReduser.suitableMasters,
      currentOrder: state.clientOrderReduser.currentOrder,
      masterListIsLoad: state.clientServices.masterListIsLoad,
      isAuth: state.clientServices.isAuth,
    };
  });
  const dispatch = useDispatch();

  function submitHandler(e) {
    e.preventDefault();
    let masterId = e.target.chooseMaster.value;
    if (!e.target.chooseMaster.length) {
      if (!e.target.chooseMaster.checked) {
        masterId = null;
      }
    }
    if (Object.keys(state.currentOrder).length === 1) {
      return false;
    }
    let endOrderTime;
    let clientTimeHour = +state.currentOrder.time.match(/[^:]+/);
    let clientTimeMin = state.currentOrder.time.match(/:\d\d$/);
    switch (state.currentOrder.size) {
      case "small":
        endOrderTime = clientTimeHour + 1 + clientTimeMin;
        break;
      case "middle":
        endOrderTime = clientTimeHour + 2 + clientTimeMin;
        break;
      case "large":
        endOrderTime = clientTimeHour + 3 + clientTimeMin;
        break;
      default:
        endOrderTime = 0;
    }
    let newObj = {
      ...state.currentOrder,
      masterId: masterId,
      endTime: endOrderTime,
    };

    dispatch(changeMasterListIsLoad(true));
    fetch(`${SERVERDOMAIN}/orders/post`, {
      method: "POST",
      headers: {
        Authorization: localStorage.getItem("token")
          ? "Bearer " + localStorage.getItem("token")
          : "",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(newObj),
    })
      .then((json) => json.json())
      .then((data) => {
        dispatch(changeMasterListIsLoad(false));
        if (data.success) {
          sendMail(`${SERVERDOMAIN}/send_message`, data.payload);
          dispatch(
            addCurrentOrderToState(
              state.isAuth
                ? {
                    email: JSON.parse(localStorage.getItem("user")).email,
                  }
                : {}
            )
          );
          dispatch(
            changeSuccessModalData({
              msg: data.msg,
              backBtnTxt: "Back to order form",
              backTo: "/",
            })
          );
          document.getElementById("callSuccessModalBtn").click();
        } else {
          dispatch(
            changeWarningModalData({
              msg: data.msg,
            })
          );
          document.getElementById("callWarningModalBtn").click();
        }
      })
      .catch((err) => {
        dispatch(changeMasterListIsLoad(false));
        alert(err);
      });
  }
  return (
    <FreeMastersForm
      submitHandler={submitHandler}
      suitableMasters={state.suitableMasters}
      isLoad={state.masterListIsLoad}
      backTo={"/client"}
      isMakeOrder={Object.keys(state.currentOrder).length <= 1 ? true : false}
    />
  );
}

MastersList.propTypes = {
  mastersArr: PropTypes.array,
  submitHandler: PropTypes.func,
  history: PropTypes.object,
  isAuth: PropTypes.bool,
};

export default MastersList;
