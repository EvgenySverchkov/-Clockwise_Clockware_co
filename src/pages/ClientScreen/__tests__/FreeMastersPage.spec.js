import React from "react";
import {create, act} from "react-test-renderer";
import configMockStore from "redux-mock-store";
import {Provider} from "react-redux";
import {MemoryRouter, Route} from "react-router-dom"

import FreeMastersPage from "../FreeMastersPage";
import Context from "../../../ContextComponent";

const mockStore = configMockStore();

describe("Test of <FreeMastersPage />", ()=>{
    it("Default view", ()=>{
        const store = mockStore({
            masterReducer: {
                suitableMasters: [{id: 1, rating: 5, name: "Name"}]
            },
            ordersReducer: {
                currentOrder: {name:"Alex", email:"199816a@mail.ru"},
                orderFormIsLoad: false,
            },
            authReducer: {
                isAuth: false
            }
        });
        const mockProps = {
            history: {
                push: jest.fn()
            }
        }
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <MemoryRouter initialEntries={['/']}>
                        <Route path="/">
                            <FreeMastersPage {...mockProps}/>
                        </Route>
                    </MemoryRouter>
                </Context.Provider>
            </Provider>
        );
    
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("Call submit event when master was choosed and server return object with success field with true value", async ()=>{
        global.fetch = jest.fn(()=>Promise.resolve({json: jest.fn(()=>Promise.resolve({success: true}))}));
        const store = mockStore({
            masterReducer: {
                suitableMasters: [{id: 1, rating: 5, name: "Name"}]
            },
            ordersReducer: {
                currentOrder: {name:"Alex", email:"199816a@mail.ru", time: "15:00"},
                orderFormIsLoad: false,
            },
            authReducer: {
                isAuth: false
            }
        });
        const mockProps = {
            history: {
                push: jest.fn()
            }
        }
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                chooseMaster: {
                    value: 1
                }
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <MemoryRouter initialEntries={['/']}>
                        <Route path="/">
                            <FreeMastersPage {...mockProps}/>
                        </Route>
                    </MemoryRouter>
                </Context.Provider>
            </Provider>
        );
        await act(async ()=>{
            await component.root.findByType("form").props.onSubmit(eventObject)
        });
        expect(mockContextValue.openSuccessWindowWithMsg.mock.calls.length).toBe(1);
        expect(mockProps.history.push.mock.calls.length).toBe(1);
    });
    it("Call submit event when master was choosed and server return object with success field with false value", async ()=>{
        global.fetch = jest.fn(()=>Promise.resolve({json: jest.fn(()=>Promise.resolve({success: false}))}));
        const store = mockStore({
            masterReducer: {
                suitableMasters: [{id: 1, rating: 5, name: "Name"}]
            },
            ordersReducer: {
                currentOrder: {name:"Alex", email:"199816a@mail.ru", time: "15:00"},
                orderFormIsLoad: false,
            },
            authReducer: {
                isAuth: false
            }
        });
        const mockProps = {
            history: {
                push: jest.fn()
            }
        }
        const mockContextValue = {
            openSuccessWindowWithMsg: jest.fn(),
            openErrorWindowWithMsg: jest.fn(),
            openWarningTooltip: jest.fn(),
            warningTooltipMsg: "Warning"
        }
        const eventObject = {
            preventDefault: jest.fn(),
            target: {
                chooseMaster: {
                    value: 1
                }
            }
        }
        let component = create(
            <Provider store={store}>
                <Context.Provider value = {mockContextValue}>
                    <MemoryRouter initialEntries={['/']}>
                        <Route path="/">
                            <FreeMastersPage {...mockProps}/>
                        </Route>
                    </MemoryRouter>
                </Context.Provider>
            </Provider>
        );
        await act(async ()=>{
            await component.root.findByType("form").props.onSubmit(eventObject)
        });
        expect(mockContextValue.openErrorWindowWithMsg.mock.calls.length).toBe(1);
    });
});