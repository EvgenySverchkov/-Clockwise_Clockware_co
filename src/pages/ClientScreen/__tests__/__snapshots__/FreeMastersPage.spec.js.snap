// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <FreeMastersPage /> 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-sm-4 mb-2 col-sm-6 col-lg-4"
    >
      <div
        className="card"
      >
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            Name
          </h5>
          <p
            className="card-text"
          >
            Rating: 
            5
          </p>
          <div
            className="form-check form-check-inline"
          >
            <input
              className="form-check-input"
              id={1}
              name="chooseMaster"
              onChange={null}
              type="radio"
              value={1}
            />
            <label
              className="form-check-label"
              htmlFor={1}
            >
              Choose
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="form-group float-left mt-3"
  >
    <a
      className="btn btn-primary"
      href="/client"
      onClick={[Function]}
    >
      Ð¡ome back
    </a>
  </div>
  <div
    className="form-group float-right mt-3"
  >
    <input
      className="btn btn-primary"
      type="submit"
      value="Book now"
    />
  </div>
</form>
`;
